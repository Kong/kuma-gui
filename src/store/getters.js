export const workspace = (state) => state.workspace

export const getSelectedMesh = (state) => {
  return state.selectedMesh
}
