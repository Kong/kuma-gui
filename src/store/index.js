import Vue from 'vue'
import Vuex from 'vuex'

import sidebar from '@/store/modules/sidebar'
// import workspaces from '@/store/modules/workspaces'

Vue.use(Vuex)

export default (api) => {
  const store = new Vuex.Store({
    modules: {
      sidebar
      // workspaces
    },
    state: {
      onboardingComplete: false,
      globalLoading: null,
      meshes: [],
      dataplanes: [],
      selectedMesh: 'default',
      totalMeshCount: 0,
      totalDataplaneCount: 0,
      totalDataplaneList: [],
      anyDataplanesOffline: null,
      totalDataplaneCountFromMesh: 0,
      totalTrafficRoutesCountFromMesh: 0,
      totalTrafficPermissionsCountFromMesh: 0,
      totalTrafficLogsCountFromMesh: 0,
      tagline: null,
      version: null,
      status: null
    },
    getters: {
      getOnboardingStatus (state) {
        return state.onboardingComplete
      },
      globalLoading (state) {
        return state.globalLoading
      },
      getSelectedMesh (state) {
        return state.selectedMesh
      },
      getMeshList (state) {
        return state.meshes
      },
      getDataplanes (state) {
        return state.dataplanes
      },
      getDataplanesList (state) {
        return state.totalDataplaneList
      },
      getAnyDpOffline (state) {
        return state.anyDataplanesOffline
      },
      getTotalMeshCount (state) {
        return state.totalMeshCount
      },
      getTotalDataplaneCount (state) {
        return state.totalDataplaneCount
      },
      getTotalDataplaneCountFromMesh (state) {
        return state.totalDataplaneCountFromMesh
      },
      getTotalTrafficRoutesCountFromMesh (state) {
        return state.totalTrafficRoutesCountFromMesh
      },
      getTotalTrafficPermissionsCountFromMesh (state) {
        return state.totalTrafficPermissionsCountFromMesh
      },
      getTrafficLogsFromMeshTotalCount (state) {
        return state.totalTrafficLogsCountFromMesh
      },
      getVersion (state) {
        return state.version
      },
      getTagline (state) {
        return state.tagline
      },
      getStatus (state) {
        return state.status
      }
    },
    mutations: {
      SET_ONBOARDING_STATUS (state, status) {
        state.onboardingComplete = status
      },
      SET_GLOBAL_LOADING (state, { globalLoading }) {
        state.globalLoading = globalLoading
      },
      FETCH_ALL_MESHES (state, meshes) {
        state.meshes = meshes
      },
      FETCH_DATAPLANES_FROM_MESH (state, dataplanes) {
        state.dataplanes = dataplanes
      },
      SET_SELECTED_MESH (state, mesh) {
        state.selectedMesh = mesh
      },
      SET_TOTAL_MESH_COUNT (state, count) {
        state.totalMeshCount = count
      },
      SET_TOTAL_DP_COUNT (state, count) {
        state.totalDataplaneCount = count
      },
      SET_TOTAL_DP_LIST (state, dataplanes) {
        state.totalDataplaneList = dataplanes
      },
      SET_ANY_DP_OFFLINE (state, status) {
        state.anyDataplanesOffline = status
      },
      SET_TOTAL_DP_COUNT_FROM_MESH (state, count) {
        state.totalDataplaneCountFromMesh = count
      },
      SET_TOTAL_TRAFFIC_ROUTES_COUNT_FROM_MESH (state, count) {
        state.totalTrafficRoutesCountFromMesh = count
      },
      SET_TOTAL_TRAFFIC_PERMISSIONS_COUNT_FROM_MESH (state, count) {
        state.totalTrafficPermissionsCountFromMesh = count
      },
      SET_TOTAL_TRAFFIC_LOGS_COUNT_FROM_MESH (state, count) {
        state.totalTrafficLogsCountFromMesh = count
      },
      SET_VERSION (state, version) {
        state.version = version
      },
      SET_TAGLINE (state, tagline) {
        state.tagline = tagline
      },
      SET_STATUS (state, status) {
        state.status = status
      }
    },
    actions: {
      // update the onboarding state
      updateOnboardingStatus ({ commit }, status) {
        commit('SET_ONBOARDING_STATUS', status)
      },

      // fetch all of the meshes from the API
      fetchMeshList ({ commit }) {
        return api.getAllMeshes()
          .then(response => {
            commit('FETCH_ALL_MESHES', response)
          })
          .catch(error => {
            console.error(error)
          })
      },

      // fetch all dataplanes from a specific mesh
      fetchDataplanesFromMesh ({ commit }, mesh) {
        return api.getAllDataplanesFromMesh(mesh)
          .then(response => {
            commit('FETCH_DATAPLANES_FROM_MESH', response)
          })
          .catch(error => {
            console.error(error)
          })
      },

      // update the selected mesh
      updateSelectedMesh ({ commit }, mesh) {
        commit('SET_SELECTED_MESH', mesh)
      },

      // get the total number of meshes
      getMeshTotalCount ({ commit }) {
        return api.getAllMeshes()
          .then(response => {
            const total = response.items.length

            commit('SET_TOTAL_MESH_COUNT', total)
          })
          .catch(error => {
            console.error(error)
          })
      },

      // a makeshift way to get the total amount of dataplanes present
      getDataplaneTotalCount ({ commit }) {
        const getDataplanes = async () => {
          const meshes = await api.getAllMeshes()
          const result = []

          for (let i = 0; i < meshes.items.length; i++) {
            const dataplanes = await api.getAllDataplanesFromMesh(meshes.items[i].name)
            const dpCount = await dataplanes.items.length

            result.push(dpCount)
          }

          const reduced = result.reduce((a, b) => a + b, 0)

          commit('SET_TOTAL_DP_COUNT', reduced)
        }

        getDataplanes()
      },

      // a makeshift way to get the list of all present dataplanes across all
      // meshes this will also set a status for whether or not any of the
      // dataplanes are offline
      getAllDataplanes ({ commit }) {
        const getDataplanes = async () => {
          return new Promise(async (resolve, reject) => {
            const meshes = await api.getAllMeshes()
            const result = []
            const states = []

            // create a status label column
            for (let i = 0; i < meshes.items.length; i++) {
              const mesh = meshes.items[i].name
              const dataplanes = await api.getAllDataplanesFromMesh(mesh)
              const items = await dataplanes.items

              for (let i = 0; i < items.length; i++) {
                const itemName = items[i].name
                const itemMesh = items[i].mesh

                const itemStatus = await api.getDataplaneOverviews(mesh, itemName)
                  .then(response => {
                    const items = response.dataplaneInsight.subscriptions

                    if (items && items.length > 0) {
                      for (let i = 0; i < items.length; i++) {
                        const connectTime = items[i].connectTime
                        const disconnectTime = items[i].disconnectTime

                        if (connectTime && connectTime.length && !disconnectTime) {
                          return 'Online'
                        } else {
                          return 'Offline'
                        }
                      }
                    } else {
                      return 'Offline'
                    }
                  })

                // create the full data array
                result.push({
                  status: itemStatus,
                  name: itemName,
                  mesh: itemMesh
                })
              }
            }

            // create a simple flat status object with booleans for checking
            // if any // dataplanes are offline
            for (let i = 0; i < Object.values(result).length; i++) {
              const statusVal = Object.values(result[i])[0]
              const isOnline = !(statusVal === 'Offline' || statusVal === 'offline')

              states.push(isOnline)
            }

            // if any of the dataplanes return false for being online
            // commit this so we can check against it
            const anyDpOffline = states.some(i => i === false)

            commit('SET_ANY_DP_OFFLINE', anyDpOffline)

            // commit the total list of dataplanes
            commit('SET_TOTAL_DP_LIST', result)

            // resolve the promise
            resolve()
          })
        }

        return getDataplanes()
      },

      // get the total number of dataplanes from a mesh
      getDataplaneFromMeshTotalCount ({ commit }, mesh) {
        return api.getAllDataplanesFromMesh(mesh)
          .then(response => {
            const total = response.items.length

            commit('SET_TOTAL_DP_COUNT_FROM_MESH', total)
          })
          .catch(error => {
            console.error(error)
          })
      },

      // get the total number of traffic routes from a mesh
      getTrafficRoutesFromMeshTotalCount ({ commit }, mesh) {
        return api.getTrafficRoutes(mesh)
          .then(response => {
            const total = response.items.length

            commit('SET_TOTAL_TRAFFIC_ROUTES_COUNT_FROM_MESH', total)
          })
          .catch(error => {
            console.error(error)
          })
      },

      // get the total number of traffic permissions from a mesh
      getTrafficPermissionsFromMeshTotalCount ({ commit }, mesh) {
        return api.getTrafficPermissions(mesh)
          .then(response => {
            const total = response.items.length

            commit('SET_TOTAL_TRAFFIC_PERMISSIONS_COUNT_FROM_MESH', total)
          })
          .catch(error => {
            console.error(error)
          })
      },

      // get the total number of traffic logs from a mesh
      getTrafficLogsFromMeshTotalCount ({ commit }, mesh) {
        return api.getTrafficLogs(mesh)
          .then(response => {
            const total = response.items.length

            commit('SET_TOTAL_TRAFFIC_LOGS_COUNT_FROM_MESH', total)
          })
          .catch(error => {
            console.error(error)
          })
      },

      // get the current version
      getVersion ({ commit }) {
        return api.getInfo()
          .then(response => {
            commit('SET_VERSION', response.version)
          })
          .catch(error => {
            console.error(error)
          })
      },

      // get the current tagline
      getTagline ({ commit }) {
        return api.getInfo()
          .then(response => {
            commit('SET_TAGLINE', response.tagline)
          })
          .catch(error => {
            console.error(error)
          })
      },

      // get the status of the API
      getStatus ({ commit }) {
        return api.getStatus()
          .then(response => {
            commit('SET_STATUS', response.statusText)
          })
      }
    }
  })

  return store
}
