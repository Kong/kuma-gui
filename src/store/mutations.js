export default {
  FETCH_MESHES (state, meshes) {
    state.meshes = meshes
  }
}
