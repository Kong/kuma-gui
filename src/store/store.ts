import { InjectionKey } from 'vue'
import { createStore, Store } from 'vuex'

import { storeConfig, State } from './index'

export const storeKey: InjectionKey<Store<State>> = Symbol('store')

export const store = createStore<State>(storeConfig)
