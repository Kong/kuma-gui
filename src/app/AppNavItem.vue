<template>
  <div
    class="nav-item"
    :class="[
      { 'nav-item--is-active': isActive },
      { 'nav-item--is-link': targetRoute !== null },
    ]"
    :data-testid="link"
  >
    <router-link
      v-if="targetRoute !== null"
      class="nav-item__link"
      :to="targetRoute"
      @click="onNavItemClick"
    >
      {{ name }}

      <span
        v-if="amount"
        class="amount"
        :class="{ 'amount--empty': amount === '0' }"
      >
        {{ amount }}
      </span>
    </router-link>

    <div
      v-else
      class="nav-item__title"
    >
      <span class="text-uppercase">
        {{ name }}
      </span>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { computed } from 'vue'
import { useRoute, useRouter, RouteLocationNamedRaw } from 'vue-router'
import { datadogLogs } from '@datadog/browser-logs'
import { datadogLogEvents } from '@/datadogEvents'

import { useStore } from '@/store/store'
import { get } from '@/utils/get'

const route = useRoute()
const router = useRouter()
const store = useStore()

const props = defineProps({
  link: {
    type: String,
    required: false,
    default: '',
  },

  insightsFieldAccessor: {
    type: String,
    required: false,
    default: '',
  },

  name: {
    type: String,
    required: false,
    default: '',
  },

  usesMeshParam: {
    type: Boolean,
    required: false,
    default: false,
  },
})

const amount = computed(() => {
  if (props.insightsFieldAccessor) {
    const value = get(store.state.sidebar.insights, props.insightsFieldAccessor, 0)
    return value > 99 ? '99+' : String(value)
  } else {
    return ''
  }
})

const targetRoute = computed<RouteLocationNamedRaw | null>(() => {
  if (props.link === '') {
    return null
  }

  const targetRoute: RouteLocationNamedRaw = {
    name: props.link,
  }

  // Sets `mesh` params only if route actually has `mesh` param defined.
  // See: https://github.com/vuejs/router/blob/main/packages/router/CHANGELOG.md#414-2022-08-22
  if (props.usesMeshParam) {
    targetRoute.params = {
      mesh: store.state.selectedMesh,
    }
  }

  return targetRoute
})

const isActive = computed(() => {
  if (targetRoute.value === null) {
    return false
  }

  if (props.link === route.name) {
    return true
  }

  const currentRouteSubpath = route.path.split('/')[2]
  if (currentRouteSubpath === targetRoute.value.name) {
    return true
  }

  if (route.meta.parent) {
    try {
      const parentRoute = router.resolve({ name: route.meta.parent })

      if (parentRoute.name === props.link) {
        return true
      }
    } catch (error) {
      if (error instanceof Error && error.message.includes('No match for')) {
        // Unfortunately, `router.resolve` throws instead of returning `null` when a route can’t be resolved so we ignore this particular error because we don’t care about non-existing routes here.
        console.warn(error)
      } else {
        throw error
      }
    }
  }

  return props.link && route.matched.some((r) => props.link === r.name || props.link === r.redirect)
})

function onNavItemClick() {
  datadogLogs.logger.info(datadogLogEvents.SIDEBAR_ITEM_CLICKED, { data: targetRoute.value })
}
</script>

<style lang="scss" scoped>
.nav-item {
  position: relative;
  display: flex;
  white-space: nowrap;
  overflow: hidden;
  margin-left: var(--spacing-xs);
  margin-bottom: var(--spacing-xxs);
  border-radius: 5px;
}

.nav-item--is-active {
  font-weight: 500;
  background-color: var(--SidebarLinkBGColor);
}

.nav-item--is-link:hover {
  background: var(--SidebarLinkBGColor);
}

.nav-item__title {
  margin-left: var(--spacing-xs);
  padding-top: var(--spacing-xs);
  padding-bottom: var(--spacing-xs);
  font-weight: 500;
  font-size: var(--type-sm);
  color: var(--SidebarTitleColor);
}

.nav-item__link {
  display: flex;
  width: 100%;
  align-items: center;
  padding: var(--spacing-xs) var(--spacing-md);
  text-decoration: none;
  color: var(--SidebarLinkColor);
}

.amount {
  position: absolute;
  top: 0;
  right: 8px;
  bottom: 0;
  width: 1.5rem;
  height: 1.25rem;
  margin: auto;
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid #fff;
  border-radius: 0.25rem;
  font-size: 0.75rem;
  font-weight: 400;
  background-color: var(--gray-2);
}

.amount--empty {
  background-color: var(--gray-4);
}
</style>
