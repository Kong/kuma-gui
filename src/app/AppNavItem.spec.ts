import { createRouter, createWebHashHistory } from 'vue-router'
import { mount } from '@vue/test-utils'

import AppNavItem from './AppNavItem.vue'
import { store, storeKey } from '@/store/store'

const router = createRouter({
  history: createWebHashHistory(),
  routes: [
    {
      path: '/',
      name: 'home',
      component: { template: 'TestComponent' },
    },
    {
      path: '/mesh/:mesh/data-planes',
      name: 'data-plane-list-view',
      component: { template: 'TestComponent' },
    },
  ],
})

function renderComponent() {
  return mount(AppNavItem, {
    props: {
      name: 'All',
      routeName: 'data-plane-list-view',
      usesMeshParam: true,
      insightsFieldAccessor: 'mesh.dataplanes.total',
    },
    global: {
      plugins: [router, [store, storeKey]],
    },
  })
}

describe('AppNavItem.vue', () => {
  it('renders snapshot with link to selected mesh', () => {
    const wrapper = renderComponent()

    expect(wrapper.element).toMatchSnapshot()
  })

  it.each([
    [0, '0'],
    [20, '20'],
    [200, '99+'],
  ])('renders amount of items', (numberOfDataplanes, expectedText) => {
    store.state.sidebar.insights.mesh.dataplanes.total = numberOfDataplanes

    const wrapper = renderComponent()

    expect(wrapper.html()).toContain(expectedText)
  })
})
