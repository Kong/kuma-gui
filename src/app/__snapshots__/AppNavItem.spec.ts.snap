// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppNavItem.vue renders snapshot with link to selected mesh 1`] = `
<div
  class="nav-item nav-item--is-link"
  data-testid="data-plane-list-view"
>
  <a
    class="nav-item__link"
    href="#/mesh/all/data-planes"
  >
    All 
    <span
      class="amount amount--empty"
    >
      0
    </span>
  </a>
</div>
`;
